Dev/2.0 metastore readme (#451)
ray-project/deltacat: deltacat/storage/README.md

> # Deltacat Storage Model Specification (WORK IN PROGRESS)

Author: woah, TIL about github suggested changes. It's super convenent

This document describes the **Deltacat Metastore Model**, which manages a 
hierarchical set of objects:

- **Namespace**
- **Table**
- **Table Version**
- **Stream**
- **Partition**
- **Delta**

## Naming, Mutability, and Canonical Strings

> All objects have an "id" field. For Table Versions, Streams, Deltas, and Partitions with "named immutable ids", the user provided id sets the "id" field. 

Reviewer: ```suggestion
All objects have an "id" field. For Table Versions, Streams, Deltas, and Partitions with "named immutable ids", the 'id' field must be manually assigned prior to commit.
```

> All objects have an "id" field. For Table Versions, Streams, Deltas, and Partitions with "named immutable ids", the user provided id sets the "id" field. 

Author: Ok, sounds like there is still a missing storage API layer of abstraction to dictate e.g. if the table version name is directly from user input or created by storage api impl. Will update

| **Table Version**   | `name`                                  | Immutable       |
| **Stream**          | `stream_id` + `format`                  | Immutable       |
| **Partition**       | `partition_id` + `partition_values`     | Immutable       |
| **Delta**           | `stream_position`                       | Immutable    

# Directory/file structure
## Object Directories (Immutable id)
Every metastore object (Namespace, Table, TableVersion, etc.) has a root 
directory. 

named by its **immutable ID** (e.g., a UUID or a numeric ID). 
Inside that directory:

For instance, a namespace root looks like:
```
${catalog_root}/${namespace_id}/
    └──rev/
        ...
    └──${id-table-1}/
        ...
    └──${id-table-2}/
        ...
```
**Child object directories**

The parent object has a single directory for each child object directory. Those directories in turn use the structure described above.

**Revision Directory**
The **`rev/`** directory contains versioned metadata files named:
`<revision_number_padded_20_digits>_<txn_operation_type>_<txn_id>.mpk`

- `revision_number_padded_20_digits` is zero-padded, e.g., `00000000000000000001`.

> - `txn_operation_type` is typically `create`, `update`, or `delete`.

Reviewer: ```suggestion
- `txn_operation_type` is the type of operation a transaction applied to the object, which is either `create`, `update`, or `delete`.
```

> - `txn_id` is a unique identifier for the transaction that performed this operation.

Reviewer: ```suggestion
- `txn_id` is the unique id of the transaction that created this revision file.
```

### Mutable Name Directories
Certain objects (like **Namespaces** and **Tables**) may have **mutable** names. 

> To support renames while keeping an **immutable** ID, the code can create
> a “digest-based” directory:

Reviewer: ```suggestion
To support object renames and alias name creation while keeping an **immutable** ID, we create
a "Name Resolution Directory" to map the object's mutable name or alias back to its immutable ID, with the following properties:
```

> - The **digest** is a SHA-1 hash of the object’s **canonical locator** string 

Reviewer: ```suggestion
- The name of the directory is a SHA-1 **digest** of the **canonical string** of the associated object Locator. 
```

>   (e.g., `"MyNamespace"` or `"MyNamespace|MyTable"`).

Reviewer: ```suggestion
  (e.g., `sha1_hexdigest("MyNamespace")` or `sha1_hexdigest("MyNamespace|MyTable")`).
```

> - This directory name is simply that SHA-1 digest (hex-encoded).

Reviewer: ```suggestion

```

> - Inside this **mutable name directory**, there is a single **Name Mapping File** 

Reviewer: ```suggestion
- Inside this **Name Resolution Directory**, there is a single **Name Mapping File** 
```

>   (zero bytes) that references the **immutable ID** directory.

Reviewer: ```suggestion
  (zero bytes) that references the **Immutable ID** directory.
```

**Name Mapping File**

> The mutable name directory just contains a file which maps to its immutable directory. The format of this file is: <revision_number_padded_20_digits>_<txn_operation_type>_<txn_id>.<object_id>

Reviewer: ```suggestion
The mutable name directory just contains a file which maps to its immutable directory. The format of this file is: <revision_number_padded_20_digits>_<txn_operation_type>_<txn_id>.<object_id>
Here, `object_id` is the name of the associated object's **Immutable ID** directory. Note that (except immutable ID) this is the same format used by metadata revision files, and the same process is employed to `create`, `update`, and `delete` name mappings.
```

### Transaction Directory

> The transaction directory (/txn) is a special directory in the catalog root which holds all successfully committed transactions. It contains one empty file per successful transaction.

Reviewer: Transaction files will be populated in my next PR, so may as well update this now:

```suggestion
The transaction directory (${catalog_root}/txn) is a special directory in the catalog root which holds all successfully committed transactions. It contains one file per successful transaction, whose contents record details about all operations fulfilled by the transaction.
```

```
${CATALOG_ROOT}/txn/
  |-{tx-1-id}
  |-{tx-2-id}
```
## Aliases

> TODO

Reviewer: I've described them a bit above in the context of name -> immutable ID mapping, and I haven't yet finalized any code yet to allow general assignment of N aliases to an immutable ID (the current code only allows for assignment of a single alternative alias back to the immutable ID), so further discussion can probably wait.

Author: sure will remove

## Example 
Putting all of the above together, here is an example catalog directory with one namespace and one table
```
${CATALOG_ROOT}/txn/                        # Global TXN logs

>    └──20444dca-7e3d-4dd2-af70-23b8bc890966  # Empty tx file (GUID) 

Reviewer: ```suggestion
   └──1737070822189-20444dca-7e3d-4dd2-af70-23b8bc890966  # txn log file (<txn-start-time>-<txn-uuid>)
```

> ${CATALOG_ROOT}/<namespace_digest>/         # Mutable name directory for namespace. SHA-1 of "MyNamespace"

Reviewer: ```suggestion
${CATALOG_ROOT}/<namespace_digest>/         # Mutable Name Resolution Directory for Namespace (SHA-1 Hex Digest of "MyNamespace")
```
